internal/migration/migration.go

```diff
@@ -0,0 +1,478 @@
+	"github.com/drand/drand/log"
+)
+
+type (
+	beacon struct {
+		PreviousSig []byte `json:",omitempty"`
+		Round       uint64
+		Signature   []byte
+	}
+
+	migrator struct {
+		ctx       context.Context
+		ctxCancel context.CancelFunc
+
+		startedAt time.Time
+		logger    log.Logger
+
+		pgDSN            string
+		bufferSize       int
+		beaconName       string
+		sourceBeaconPath string
+
+		// existingDB   *bbolt.DB
+		existingRows int
+
+		distChan    chan beacon
+		errChan     chan error
+		destination chain.StorageType
+	}
+)
+
+const ownerOnly = 0600
+
+const DefaultBufferSize = 10_000
+
+var (
+	// ErrMigrationNotNeeded ...
+	ErrMigrationNotNeeded = fmt.Errorf("migration not needed")
+
+	bucketName = []byte("beacons")
+)
+
+//nolint:lll // This function uses the right amount of params
+func Migrate(ctx context.Context, logger log.Logger, sourceBeaconPath, beaconName string, destination chain.StorageType, pgDSN string, bufferSize int) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	startedAt := time.Now()
+
+	if err := shouldMigrate(ctx, logger, sourceBeaconPath, beaconName, destination, pgDSN); err != nil {
+		logger.Warnw("decided storage format migration is not needed", "err", err)
+		if errors.Is(err, ErrMigrationNotNeeded) {
+			return err
+		}
+
+		return nil
+	}
+
+	bufferSize, err := computerBufferSize(bufferSize, logger, sourceBeaconPath)
+	if err != nil {
+		return err
+	}
+
+	m := migrator{
+		logger:    logger,
+		startedAt: startedAt,
+
+		pgDSN:            pgDSN,
+		bufferSize:       bufferSize,
+		beaconName:       beaconName,
+		sourceBeaconPath: sourceBeaconPath,
+		destination:      destination,
+
+		ctx:       ctx,
+		ctxCancel: cancel,
+
+		distChan: make(chan beacon, bufferSize),
+
+		errChan: make(chan error),
+	}
+
+	return m.doMigrate(ctx)
+}
+
+func computerBufferSize(bufferSize int, logger log.Logger, sourceBeaconPath string) (int, error) {
+	switch {
+	case bufferSize < 0:
+		logger.Infow("buffer size not specified, defaulting to 10000")
+		bufferSize = DefaultBufferSize
+	case bufferSize == 0:
+		var err error
+		bufferSize, err = automaticBufferSize(logger, sourceBeaconPath)
+		if err != nil {
+			return 0, err
+		}
+	case bufferSize < DefaultBufferSize:
+		logger.Warnw("buffer size seems a bit too small. The migration process might be slow", "bufferSize", bufferSize)
+	//nolint:gomnd // See below
+	case bufferSize > 10_000_000:
+		//nolint:lll // This line has the right amount of chars
+		logger.Warnw("buffer size seems a bit too large. Make sure your system can allocate enough system memory for this", "bufferSize", bufferSize)
+	}
+	return bufferSize, nil
+}
+
+func automaticBufferSize(logger log.Logger, sourceBeaconPath string) (int, error) {
+	logger.Warnw("buffer size set to 0. Running automatic buffer inference. Make sure you have enough system memory for this")
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return -1, err
+	}
+	defer func() {
```

Comment:
Bummer we need the literal function since Close() returns an error. We don't need the use of the blank identifier.